File: backend/src/main/java/com/bookbook/domain/user/repository/UserRepository.java
package com.bookbook.domain.user.repository;

import com.bookbook.domain.user.entity.User;
import com.bookbook.domain.user.enums.UserStatus;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;


public interface UserRepository extends JpaRepository<User, Long> {
    Optional<User> findByNickname(String nickname);
    Optional<User> findByUsername(String username);
    Optional<User> findByEmail(String email);
    List<User> findAllByUserStatusAndResumedAtBefore(UserStatus userStatus, LocalDateTime now);

    @Query("""
        SELECT u FROM User u WHERE
        (:status IS NULL OR u.userStatus in :status) AND
        (:userId IS NULL OR u.id = :userId)
        ORDER BY u.createAt DESC
    """)
    Page<User> findFilteredUsers(
            Pageable pageable,
            @Param("status") List<UserStatus> status,
            @Param("userId") Long userId
    );

    boolean existsByUsername(String username); // 사용자명 중복확인
    boolean existsByNickname(String nickname); // 닉네임 중복확인
}

File: backend/src/main/java/com/bookbook/domain/user/repository/UserRepository.kt
package com.bookbook.domain.user.repository

import com.bookbook.domain.user.entity.User
import com.bookbook.domain.user.enums.UserStatus
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param
import java.util.*

interface UserRepository : JpaRepository<User, Long> {
    fun findByUsername(username: String): Optional<User>

    @Query("""
        SELECT u FROM User u WHERE 
        u.userStatus = com.bookbook.domain.user.enums.UserStatus.SUSPENDED AND 
        u.resumedAt IS NOT NULL AND
        u.resumedAt <= NOW() ORDER BY 
        u.resumedAt ASC LIMIT 100
    """)
    fun findAllPossibleResumedUsers(): List<User>

    @Query(
        """
        SELECT u FROM User u WHERE
        (:status IS NULL OR u.userStatus in :status) AND
        (:userId IS NULL OR u.id = :userId)
        ORDER BY u.createdDate DESC
    """
    )
    fun findFilteredUsers(
        pageable: Pageable?,
        @Param("status") status: List<UserStatus>?,
        @Param("userId") userId: Long?
    ): Page<User>

    fun existsByUsername(username: String): Boolean
    fun existsByNickname(nickname: String): Boolean
}